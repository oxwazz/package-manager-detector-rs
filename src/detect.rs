use regex::Regex;
use serde::Deserialize;
use std::fs;
use std::path::Path;

const AGENTS: [&str; 7] = ["npm", "yarn", "yarn@berry", "pnpm", "pnpm@6", "bun", "deno"];

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PackageJson {
    package_manager: String,
}

#[derive(Debug)]
pub struct HandlePackageManagerReturn {
    name: String,
    agent: String,
    version: String,
}

pub fn handle_package_manager(
    filepath: &Path,
    on_unknown: Option<Box<dyn Fn()>>,
) -> Option<HandlePackageManagerReturn> {
    let pkg = fs::File::open(filepath).expect("package.json failed to open");
    let pkg: Option<PackageJson> = serde_json::from_reader(pkg).ok();
    match pkg {
        None => None,
        Some(pkg_) => {
            let p = Regex::new(r"^\^").unwrap();
            let p = p.replace_all(pkg_.package_manager.as_str(), "");
            let p = p.split("@").collect::<Vec<&str>>();
            let name = p[0];
            let version = p[1];
            let major_version = version.split(".").collect::<Vec<&str>>()[0]
                .parse::<i8>()
                .unwrap_or(0);

            if name == "yarn" && major_version > 1 {
                Some(HandlePackageManagerReturn {
                    name: name.to_string(),
                    agent: String::from("yarn@berry"),
                    version: String::from("berry"),
                })
            } else if name == "pnpm" && major_version < 7 {
                Some(HandlePackageManagerReturn {
                    name: name.to_string(),
                    agent: String::from("pnpm@6"),
                    version: version.to_string(),
                })
            } else if AGENTS.contains(&name) {
                Some(HandlePackageManagerReturn {
                    name: name.to_string(),
                    agent: name.to_string(),
                    version: version.to_string(),
                })
            } else {
                match on_unknown {
                    None => None,
                    Some(cb) => {
                        cb();
                        None
                    }
                }
            }
        }
    }
}
