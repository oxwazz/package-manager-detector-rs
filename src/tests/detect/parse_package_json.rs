#[allow(unused_imports)]
use crate::detect::parse_package_json;
#[allow(unused_imports)]
use serde_json::json;
use std::io::Write;
use tempfile::NamedTempFile;

#[test]
fn test_yarn_v2_plus() {
    let content = json!({
        "packageManager": "^yarn@2.0.0"
    });
    let temp_file = create_temp_json_file(content);
    let result = parse_package_json(temp_file.path(), None);
    let package_info = result.unwrap();
    insta::assert_debug_snapshot!(package_info);
}

#[test]
fn test_yarn_v1() {
    let content = json!({
        "packageManager": "^yarn@1.22.19"
    });
    let temp_file = create_temp_json_file(content);
    let result = parse_package_json(temp_file.path(), None);
    let package_info = result.unwrap();
    insta::assert_debug_snapshot!(package_info);
}

#[test]
fn test_pnpm_v6() {
    let content = json!({
        "packageManager": "^pnpm@6.32.0"
    });
    let temp_file = create_temp_json_file(content);
    let result = parse_package_json(temp_file.path(), None);
    let package_info = result.unwrap();
    insta::assert_debug_snapshot!(package_info);
}

#[test]
fn test_pnpm_v7_plus() {
    let content = json!({
        "packageManager": "^pnpm@7.0.0"
    });
    let temp_file = create_temp_json_file(content);
    let result = parse_package_json(temp_file.path(), None);
    assert!(result.is_some());
    let package_info = result.unwrap();
    insta::assert_debug_snapshot!(package_info);
}

#[test]
fn test_other_known_agents() {
    let known_agents = ["npm", "bun"];
    for &agent in known_agents.iter() {
        let content = json!({
            "packageManager": format!("^{}@1.2.3", agent)
        });
        let temp_file = create_temp_json_file(content);
        let result = parse_package_json(temp_file.path(), None);
        let package_info = result.unwrap();
        insta::assert_debug_snapshot!(package_info);
    }
}

#[test]
fn test_unknown_agent_without_callback() {
    let content = json!({
        "packageManager": "^unknown@1.2.3"
    });
    let temp_file = create_temp_json_file(content);
    let result = parse_package_json(temp_file.path(), None);
    insta::assert_debug_snapshot!(result);
}

#[test]
fn test_unknown_agent_with_callback() {
    let mut called = false;
    let content = json!({
        "packageManager": "^unknown@1.2.3"
    });
    let temp_file = create_temp_json_file(content);
    let result = parse_package_json(
        temp_file.path(),
        Some(Box::new(|| {
            called = true;
        })),
    );
    assert!(called);
    insta::assert_debug_snapshot!(result);
}

#[test]
fn test_malformed_package_manager_string() {
    let content = json!({
        "packageManager": "malformed-string"
    });
    let temp_file = create_temp_json_file(content);
    let result = parse_package_json(temp_file.path(), None);
    insta::assert_debug_snapshot!(result);
}

// Helper function to create a temporary JSON file for testing
#[allow(dead_code)]
fn create_temp_json_file(content: serde_json::Value) -> NamedTempFile {
    let mut temp_file = NamedTempFile::new().unwrap();
    write!(
        temp_file,
        "{}",
        serde_json::to_string_pretty(&content).unwrap()
    )
    .unwrap();
    temp_file
}
