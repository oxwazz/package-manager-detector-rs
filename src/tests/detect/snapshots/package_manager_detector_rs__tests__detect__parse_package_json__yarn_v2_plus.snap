---
source: src/tests/detect/parse_package_json.rs
expression: package_info
snapshot_kind: text
---
HandlePackageManagerReturn {
    name: "yarn",
    agent: "yarn@berry",
    version: "berry",
}
